src/test/scala/io/github/spark_redshift_community/spark/redshift/RedshiftSourceSuite.scala

```diff
@@ -0,0 +1,598 @@
+      classOf[DirectMapredOutputCommitter].getName)
+    sc.hadoopConfiguration.set("fs.s3.awsAccessKeyId", "test1")
+    sc.hadoopConfiguration.set("fs.s3.awsSecretAccessKey", "test2")
+    sc.hadoopConfiguration.set("fs.s3a.access.key", "test1")
+    sc.hadoopConfiguration.set("fs.s3a.secret.key", "test2")
+
+  }
+
+  override def beforeEach(): Unit = {
+    super.beforeEach()
+    s3FileSystem = FileSystem.get(new URI(s3TempDir), sc.hadoopConfiguration)
+    testSqlContext = new SQLContext(sc)
+    expectedDataDF =
+      testSqlContext.createDataFrame(sc.parallelize(TestUtils.expectedData), TestUtils.testSchema)
+
+    // Configure a mock S3 client so that we don't hit errors when trying to access AWS in tests.
+    mockS3Client = Mockito.mock(classOf[AmazonS3Client], Mockito.RETURNS_SMART_NULLS)
+
+    when(mockS3Client.getBucketLifecycleConfiguration(anyString())).thenReturn(
+      new BucketLifecycleConfiguration().withRules(
+        new Rule().withPrefix("").withStatus(BucketLifecycleConfiguration.ENABLED)
+      ))
+
+    val mockManifest = Mockito.mock(classOf[S3Object], Mockito.RETURNS_SMART_NULLS)
+
+    when(mockManifest.getObjectContent).thenAnswer {
+      new Answer[S3ObjectInputStream] {
+        override def answer(invocationOnMock: InvocationOnMock): S3ObjectInputStream = {
+          val manifest =
+            s"""
+               | {
+               |   "entries": [
+               |     { "url": "${Utils.fixS3Url(Utils.lastTempPathGenerated)}/part-00000" }
+               |    ]
+               | }
+            """.stripMargin
+          // Write the data to the output file specified in the manifest:
+          val out = s3FileSystem.create(new Path(s"${Utils.lastTempPathGenerated}/part-00000"))
+          val ow = new OutputStreamWriter(out.getWrappedStream)
+          ow.write(unloadedData)
+          ow.close()
+          out.close()
+          val is = new ByteArrayInputStream(manifest.getBytes("UTF-8"))
+          new S3ObjectInputStream(
+            is,
+            Mockito.mock(classOf[HttpRequestBase], Mockito.RETURNS_SMART_NULLS))
+        }
+      }
+    }
+
+    when(mockS3Client.getObject(anyString(), endsWith("manifest"))).thenReturn(mockManifest)
+  }
+
+  override def afterEach(): Unit = {
+    super.afterEach()
+    testSqlContext = null
+    expectedDataDF = null
+    mockS3Client = null
+    FileSystem.closeAll()
+  }
+
+  override def afterAll(): Unit = {
+    sc.stop()
+    super.afterAll()
+  }
+
+  test("DefaultSource can load Redshift UNLOAD output to a DataFrame") {
+    // scalastyle:off
+    unloadedData =
+      """
+        |1|t|2015-07-01|1234152.12312498|1.0|42|1239012341823719|23|Unicode's樂趣|2015-07-01 00:00:00.001
+        |1|f|2015-07-02|0|0.0|42|1239012341823719|-13|asdf|2015-07-02 00:00:00.0
+        |0||2015-07-03|0.0|-1.0|4141214|1239012341823719||f|2015-07-03 00:00:00
+        |0|f||-1234152.12312498|100000.0||1239012341823719|24|___\|_123|
+        |||||||||@NULL@|
```

Comment:
Can you explain why you needed `@NULL@` here? (in the original, it was just another empty field)

Reply:
I didn't add that myself but I can see it was added in the Yelp fork (from which I forked) here: https://github.com/Yelp/spark-redshift/blob/master/src/test/scala/com/databricks/spark/redshift/RedshiftSourceSuite.scala#L153
I think we want to test how it handles @NULL@. If you `git diff` with that fork rather than the very original from databricks you might see less differences

Reply:
Oh, good point